### PR TITLE
Add a script to help managing embedded scripts externally

### PR DESCRIPTION
**Pull request by chmouel**
The idea of this script is if you have inside your script something like this :

It will replace this #include with the content of python/script.py

This make it easier to write your tasks and use linters/code checkers from your
editor on your included script.

This sparked from @afrittoli  discussion here https://github.com/tektoncd/catalog/discussions/715

Currently this is a helper so this would be up to the contributor to uses it before commit but i think we can maybe think on how to do this post merge so to be transparent.

### CODE REVIEW

File: test/embedded-script-external/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+kind: Task
+metadata:
+  name: git-clone
+spec:
+  steps:
+    - name: step1
+      script: "#include script.sh"
+```
+
+The script will see it and includes the `script.sh` in place of your "#include script.sh"
```
Comment (afrittoli) [inline test/embedded-script-external/README.md:32]: It would be nice to document that the result is sent to stdout and that if more than one YAML docs is in the source, all be parsed and separated by `---`.

File: test/embedded-script-external/embedded-script-external.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import re
+import sys
+import typing
+
+from ruamel.yaml import YAML
+
+REGEXP = r"^#include\s*([^$]*)"
+
+
+def replace(filename: str) -> typing.List:
```
Comment (afrittoli) [inline test/embedded-script-external/embedded-script-external.py:27]: Eventually we should have unit tests 🙏

**Review by afrittoli — COMMENTED** [review]
Thank you for starting this!

I think it's fine to have it under a `test` folder at least for now - eventually we might want some way to make it easier for contributors to include this in their workflow... perhaps a `make` target or a shell script.

Eventually we'll need
- unit tests for this python
- a script to trigger this and write the output in the right place
- a CI job to run the script and check that generated YAMLs are up to date

But not in this PR :)

/lgtm

File: test/embedded-script-external/README.md (hunk 1 above)
Comment (chmouel) [reply test/embedded-script-external/README.md:32]: Yep 👍🏻

File: test/embedded-script-external/embedded-script-external.py (hunk 2 above)
Comment (chmouel) [reply test/embedded-script-external/embedded-script-external.py]: Yep I added some, i didn't bother because we don't have ways to run them yet 😅

I usually do pytest but i guess [python builtin unittest](https://docs.python.org/3/library/unittest.html) should be enough for this

**Comment by chmouel** [comment]
> a script to trigger this and write the output in the right place

how would that work tho?

In main we have the "#include script.py" template files? But users (including ides/tkn hub or even curl people) would expect to be able to run the task directly from main branch, isnt it ?

maybe we can have both, we could have a comment above the `script: ` which would tell how to replace the script field to the CI jobs at post merge time, it would not matter the script content because the source of truth would be the #include script content.

Maybe that's a bit confusing, so let me rephrase my thinking : 

- Dev have task/foo/0.1/task.yaml
- task.yaml has something like this : 	
```yaml
steps:
   - name: foo
     #include foo.py
     script : | 
       #!/usr/bin/env python
       print("helo world")
```
- dev do some changes inside the foo.py and submit as PR,
- test-runner sees the #include and automatically test it with the new content from foo.py in script
- We merge the PR
- A post merge job would replace the script portion with the content from foo.py and commit back to catalog.

This allows user to have the latest task working directly and devs being able to develop into a real file for the script portion.

Let me know if that make sense 🙇🏻

**Comment by afrittoli** [comment]
> > a script to trigger this and write the output in the right place
> 
> how would that work tho?
> 
> In main we have the "#include script.py" template files? But users (including ides/tkn hub or even curl people) would expect to be able to run the task directly from main branch, isnt it ?
> 
> maybe we can have both, we could have a comment above the `script: ` which would tell how to replace the script field to the CI jobs at post merge time, it would not matter the script content because the source of truth would be the #include script content.
> 
> Maybe that's a bit confusing, so let me rephrase my thinking :
> 
> * Dev have task/foo/0.1/task.yaml
> * task.yaml has something like this :
> 
> ```yaml
> steps:
>    - name: foo
>      #include foo.py
>      script : | 
>        #!/usr/bin/env python
>        print("helo world")
> ```
> 
>  * dev do some changes inside the foo.py and submit as PR,
> * test-runner sees the #include and automatically test it with the new content from foo.py in script
> * We merge the PR
> * A post merge job would replace the script portion with the content from foo.py and commit back to catalog.

I don't like this approach too much because it means that we need two commits to do a change, and in between commits the repo is in an inconsistent state.

> 
> This allows user to have the latest task working directly and devs being able to develop into a real file for the script portion.
> 
> Let me know if that make sense 🙇🏻

The workflow I had in mind is very much similar to what we have to generated golang files in `pipeline` and other repos.

* Dev creates / updates the task in `task/foo/0.1/task.yaml.template`
* task.yaml.template has something like this :

```yaml
steps:
   - name: foo
     #include foo.py
     script : | 
       #!/usr/bin/env python
       print("helo world")
```

* Dev creates / updates the script in `task/foo/0.1/foo.py`
* Before committing, dev runs `tools/embed-scripts.sh`, which generates `task/foo/0.1/task.yaml`

* CI runs `tools/embed-scripts.sh` and diffs. If differences are found it fails. Differences could happen if:
   * if the dev updated the script and did not run the embed tool
   * if the dev made changes into the generated `task.yaml` instead of the template / script 
* `task/foo/0.1/task.yaml` alone is published to the bundle
* users only ever refer to `task/foo/0.1/task.yaml`

The paths / naming might change, we could have the template defined in a subfolder for instance, to "hide" it from users.

**Review by afrittoli — COMMENTED** [review]
I think this is a great start, thank you!
We'll need to create an infra image that embeds the script, so we can use it to run a CI job.
For that we'll need a docker file and a requirements.txt, but we can add those in a follow up.
/lgtm

**Comment by chmouel** [comment]
/remove-lifecycle stale

i didnt see your review @vinamra28 🙇🏻

**Comment by vinamra28** [comment]
/remove-lifecycle rotten
/lifecycle frozen
/reopen
@chmouel any updates on this?

**Comment by chmouel** [comment]
I don't think it's trivial to do this, this  will broke the "kuebctl apply -f https://github.com/tektoncd/catalog/path/to/task.yaml" maybe with this TEP which split the structure from the implementation we may be able to achive that but this probably need something extras on the hub end...

**Comment by chmouel** [comment]
> Not necessarily, it depends on how it's implemented. My suggestion would be:
> 
>     store in the catalog the template (YAML without the scripts), the script and the rendered version
>     the rendered version would be stored in the same location as today
>     add a CI job that renders the final task (template + script) and checks that it matches the rendered version in the repo

I think it may be a bit cumbersome for non frequent contributor, we have generation on cli for mans/golden etc.. and it's already a source of annoyances since it's easy to forget to run the command before sending the iteration. 

What about CI job commiting to the repo the final rendered version directly to a branch?

alternatively if hub could do this transparently and we say that hub is the "only" trusted source, then we would not have to store the yaml in repo.

>     provide catalog authors with a script to run to update all rendered tasks
